sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -5889,7 +5889,17 @@ class AstBuilder extends DataTypeAstBuilder
           SubqueryAlias(SubqueryAlias.generateSubqueryName(), left)
       }
       withWhereClause(c, withSubqueryAlias)
-    }.get)
+    }.getOrElse(Option(ctx.pivotClause()).map { c =>
+      if (ctx.unpivotClause() != null) {
+        throw QueryParsingErrors.unpivotWithPivotInFromClauseNotAllowedError(ctx)
+      }
+      withPivot(c, left)
+    }.getOrElse(Option(ctx.unpivotClause()).map { c =>
+      if (ctx.pivotClause() != null) {
+        throw QueryParsingErrors.unpivotWithPivotInFromClauseNotAllowedError(ctx)
```

Comment:
do we have tests to check this error?

Reply:
Good catch, no. I added some, to make sure we are returning this descriptive error message instead of a generic "syntax error".